### ISSUE TITLE
Add more granular query tracing events

### ISSUE THREAD

**Issue by phd3** [trinodb/trino#1826]
Trino emits events for query creation, query completion and split completion through the event listener. While digging into performance issues for specific queries, we often realize that more granular information about other query specific operations may be more helpful. For example, (1) events specific to Hive split discovery, (2) events corresponding to a split's scheduling, blocking, unblocking, waiting (3) events corresponding to page source operations (opening a file, reading a footer, reading data) etc. This can be extended to form a set of "tracing" events.

We want to enable this tracing for queries selectively through a session property, since the volume of data may be too much to emit for every query.

One approach to do this would be to have a "queryTracer" object inside the session, that contains references to the eventListenerManager. While creating tasks, pipelines or drivers, it'd create appropriate taskTracer, pipelineTracer, DriverTracer objects. These objects would expose methods to trace various events happening during query execution. Invocation of these methods would trigger emission of events through the event listener.  One catch here is that we want to emit events inside connectors (ConnectorSplitManager, ConnectorPageSource) which may require delegation of appropriate objects to the connector. We're looking for suggestions from the commnunity around the best way to implement this in the existing framework.

**Comment by jtao15** [comment]
### Goal
- We need more granular events for performance diagnostic purpose emitted through the event listener. We are interested in events happening throughout a query’s execution timeline. Examples include
  - Events for Hive split discovery
  - Events for split execution process (scheduling, waiting, blocking and unblocking)
  - Events corresponding to page source operations (opening a file, reading a footer, reading data) etc.
E.g. For an ORC source split, we may be interested in the time spent on reading  metadata, the timestamp when it is blocked, the block reason and the time spent reading ORC footer vs body.
- User can enable or disable this feature per query. 
- The feature is designed to offer basic diagnostic logging, and will not support for complex metrics.
### Proposed solution
- We’ll add a session attribute `query_tracing` for Presto. Users can enable this feature by setting a session property.

**Tracer**
- We will introduce tracers for Presto execution objects (query, stage, pipeline, split and operator). A tracer is a light-weighted context for the execution object and used to propagate events to the event listener from different points in query execution. 
- A tracer will keep a private reference to the event listener manager as the struct below. Once an interested action for the execution object happens, its tracer will call the event listener manager to send the corresponding event. E.g. `SplitBlocked` will be called to send a blocked event once the split is put in the blocked queue, and `SplitUnblocked` will be called when this split is removed from the blocked queue and put into the waiting queue. The tracer event is used for logging purposes, and will be sent as a payload without well-defined struct. 
```
class Tracer {
   private final EventListenerManager eventListenerManager;
   private final String queryId;
   private final Optional<String> NodeId;
   private final Optional<String> stageId;
   private final Optional<String> taskId;
   private final Optional<String> pipelineId;
   private final Optional<String> splitId;
   private final Optional<String> connectorSplitId;
   private final Optional<String> operatorId;
 
   private final void emitEvent(String payload) {
       eventListenerManager.emitEvent(queryId, "{'payload':" + payload + ",'timeNanos':'"+ System.nanoTime()+ "'}");
   }
 
  // public event methods to be invoked in the code
 
   public void CreateQuery() {
       emitEvent("{'action':'query created'}");
   }
   public void StartPlanQuery() {
       emitEvent("{'action':'query plan start'}");
   }
   .
   .
   public void SplitBlocked(String reason, long cpuNanos) { // cpu time for the executed thread before blocked             
       emitEvent("{'action':'blocked','nodeId':'" + nodeId + "','splitId':'" + splitId + "','reason':'" + reason + "','cpuTime':'" + cpuTime + "'}");
   }
   .
   .
}
```
- A tracer will be created before planning or executing of its execution object, and destroyed after it is finished. A child tracer will be created based on its parent context. 
```
Tracer queryTracer = createBasicTracer(queryId);
	Tracer taskTracer = queryTracer.withTaskId(taskId);
		Tracer splitTracer = taskTracer.withSplitId(splitId);
```
- For a certain query, a query tracer with an injection of the event listener manager will be created once a post request is received. The tracer will be passed to dispatch manager before the query is prepared. As the query is planned to create stages, each stage will have its own tracer with `splitId` copied. As the query propagates, all tasks, pipelines, splits and operators will have their own dedicated tracers with inherited contexts. The type of a certain tracer can be inferred from the id fields set. A query tracer will only have `queryId` set, a split tracer will have `queryId`, `stageId`, `splitId` set when loading, and `taskId`, `pipelineId` set before running. 

**Connector Tracer**
- For hive split discovery and page source operations, the interested actions happen in connectors. The engine tracer of the execution unit will be wrapped as a connect tracer and be passed when connector functions are called (e.g. `getSplits` for `HiveSplitManager`). The connector tracer is in charge of calling embedded engine tracer to send live events. The struct will look like below.
```
Abstract class ConnectorTracer {
    private final Tracer tracer;
 
    private final void emitEvent(String action, String payload) {
       tracer.emitEvent("{'action':'" + action + "','payload':" + payload +  "}");
   }
}
 
Public class HiveConnectorTracer
	implements ConnectorTracer
{
   public void startLoadingConnectorSplit(String connectorSplitId) {
       emitEvent("ConnectorSplitStartLoading", "{'connectorSplitId':'" + connectorSplitId + "'}");
   }
 
   public void endLoadingConnectorSplit(String connectorSplitId) {
       emitEvent("ConnectorSplitEndLoading", "{'connectorSplitId':'" + connectorSplitId + "'}");
   }
 
   public void startReadingPage() {
       emitEvent("PageStartReading", "null");
   }
}
```
### Split events we care about
With the feature enabled we can have an idea of the split lifecycle as below. 
- Connector tracers will send events of split loading related events.
- Split tracers will send all events of split actions. (E.g. `Create`, `Start`, `Schedule`, `Wait` etc). For each running period (time between `Scheduled` and `Wait`, `Block` or `Finish` shown as yellow sections), the cpu time will be included in the payload. 
- Operator actions events will be sent by operator tracers. The payload may include time spent on opening files and reading metadata for source operators
![Split_Lifecycle](https://user-images.githubusercontent.com/50065407/67589588-951f4200-f70d-11e9-880a-0b3b475166f7.jpg)
### Future work
Create an event listener plugin and save execution events in system table for easy access. We can have live output for the plugin so we can monitor time consuming query before completion as well.

**Comment by martint** [comment]
This is a great start!

A couple of things that came to mind as I was reading the proposal
* We may want events to be semi-structured. There will be a set of fields we know every event will have (e.g., a timestamp, the presto node that generated it, maybe an event type, etc). For everything else, yes, we should keep it as generic payload.
* For the payload, there are a few paths we could take:
    * Arbitrary strings. This may make downstream processing tricky.
    * JSON strings. Easier to process downstream, but limits us to a single format (what if we wanted to output more compact representations such as SMILE or other binary formats?)
    * Jackson-serializable objects. This makes it fairly trivial to serialize to JSON or not-so-trivially to other formats. On the other hand, it can be tricky to get around classloader isolation issues when trying to use custom serializers within connectors.
    * `Map<String, Object>`.  Similar to the item above, but if the kinds of objects allowed as values is constrained (simple types, arrays, maps), it's easy for the event listener to inspect, manipulate and serialize in whichever format it desires.

Other that not going with arbitrary strings, I don't yet have a strong opinion on the other alternatives. I'll continue to think about it as I digest the rest of the proposal.

**Comment by jtao15** [comment]
> This is a great start!
> 
> A couple of things that came to mind as I was reading the proposal
> 
> * We may want events to be semi-structured. There will be a set of fields we know every event will have (e.g., a timestamp, the presto node that generated it, maybe an event type, etc). For everything else, yes, we should keep it as generic payload.
> * For the payload, there are a few paths we could take:
>   
>   * Arbitrary strings. This may make downstream processing tricky.
>   * JSON strings. Easier to process downstream, but limits us to a single format (what if we wanted to output more compact representations such as SMILE or other binary formats?)
>   * Jackson-serializable objects. This makes it fairly trivial to serialize to JSON or not-so-trivially to other formats. On the other hand, it can be tricky to get around classloader isolation issues when trying to use custom serializers within connectors.
>   * `Map<String, Object>`.  Similar to the item above, but if the kinds of objects allowed as values is constrained (simple types, arrays, maps), it's easy for the event listener to inspect, manipulate and serialize in whichever format it desires.
> 
> Other that not going with arbitrary strings, I don't yet have a strong opinion on the other alternatives. I'll continue to think about it as I digest the rest of the proposal.

Thanks for the suggestion. We also want to emphasize that this feature is designed for developer debugging purposes only, the events and payload structures are tentative to change during future development. It's good to keep it agile and users should not build complex metrics based on the tracers events.
